src/file.py

```diff
@@ -0,0 +1,406 @@
+import logging
+import http
+import requests
+import cs3.storage.provider.v1beta1.resources_pb2 as cs3spr
+import cs3.storage.provider.v1beta1.provider_api_pb2 as cs3sp
+import cs3.gateway.v1beta1.gateway_api_pb2_grpc as cs3gw_grpc
+import cs3.rpc.v1beta1.code_pb2 as cs3code
+import cs3.types.v1beta1.types_pb2 as types
+import common
+
+from configparser import ConfigParser
+from typing import Iterator, Union
+from exceptions.exceptions import AuthenticationException
+from cs3resource import Resource
+from auth import Auth
+
+
+class File:
+    """
+    File class to interact with the CS3 API.
+    """
+
+    def __init__(
+        self, config: ConfigParser, log: logging.Logger, gateway: cs3gw_grpc.GatewayAPIStub, auth: Auth
+    ) -> None:
+        self.log: logging.Logger = log
+        self.config: ConfigParser = config
+        self.auth: Auth = auth
+        self.gateway: cs3gw_grpc.GatewayAPIStub = gateway
+        self.ssl_verify: bool = config.getboolean("cs3", "sslverify", fallback=True)
+        self.http_timeout: int = config.getint("cs3", "httptimeout", fallback=10)
+        self.chunk_size: int = config.getint("io", "chunksize")
+
+    def stat(self, resource: Resource) -> cs3spr.ResourceInfo:
+        """Stat a file and return the ResourceInfo object. If the file does not exist, return None."""
+        tstart = time.time()
+        res = self.gateway.Stat(
+            request=cs3sp.StatRequest(ref=resource.ref), metadata=[("x-access-token", self.auth.token)]
+        )
+        tend = time.time()
+        if res.status.code == cs3code.CODE_NOT_FOUND:
+            self.log.debug(f'msg="Invoked stat for getxattr on missing file" filepath="{resource.file_ref}"')
+            return None
+        if res.status.code == cs3code.CODE_UNAUTHENTICATED:
+            self.log.error(
+                'msg="Authentication failed on stat" filepath="%s" userid="%s" trace="%s" reason="%s"'
+                % (resource.file_ref, self.auth.token[-20:], res.status.trace, res.status.message.replace('"', "'"))
+            )
+            raise AuthenticationException(f"Failed to authenticate user {self.auth.token}, error: {res.status.message}")
+        if res.status.code != cs3code.CODE_OK:
+            self.log.error(
+                'msg="Failed to stat" filepath="%s" userid="%s" trace="%s" reason="%s"'
+                % (resource.file_ref, self.auth.token[-20:], res.status.trace, res.status.message.replace('"', "'"))
+            )
+            raise IOError(res.status.message)
+        self.log.debug(
+            f'msg="Invoked Stat" filepath="{resource.file_ref}" elapsedTimems="{(tend - tstart) * 1000:.1f}"'
+        )
+        return res.info
+
+    def set_xattr(self, resource: Resource, key: str, value: any) -> None:
+        """Set the extended attribute <key> to <value> using the given userid as access token"""
+        md = cs3spr.ArbitraryMetadata()
+        md.metadata.update({key: str(value)})  # pylint: disable=no-member
```

Comment:
I don't know why it was done this way in the wopi-server, but I will look into it, seems we could just make it a string.

Reply:
This part of the wopiserver was written with no typing at all (originally in python 2)...

Reply:
I should have phrased this better, wopiserver was using the cast, which is why I assumed it could be other types too and hence put it as "any".

Reply:
But I think we can let set it as a stringwe

Reply:
I think `string` is going to be good. The cast was needed because indeed in some cases the argument was not a string, but the caller can force the casting instead.

Reply:
>  the caller can force the casting instead.

This was my point!